src/data/equiv/embedding.lean

```diff
@@ -0,0 +1,82 @@
+/-
+Copyright (c) 2021 Eric Rodriguez. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Rodriguez
+-/
+import logic.embedding
+import data.set.lattice
+
+/-!
+# Equivalences on embeddings
+
+This file shows some advanced equivalences on embeddings, useful for constructing larger
+embeddings from smaller ones.
+-/
+
+
+open function.embedding
+
+namespace equiv
+
+/-- Embeddings from a sum type are equivalent to two separate embeddings with disjoint ranges. -/
+def sum_embedding_equiv_prod_embedding_disjoint {α β γ : Type*} :
+  ((α ⊕ β) ↪ γ) ≃ {f : (α ↪ γ) × (β ↪ γ) // disjoint (set.range f.1) (set.range f.2)} :=
+{ to_fun := λ f, ⟨(inl.trans f, inr.trans f),
+  begin
+    rintros _ ⟨⟨a, h⟩, ⟨b, rfl⟩⟩,
+    simp only [trans_apply, inl_apply, inr_apply] at h,
+    have : sum.inl a = sum.inr b := f.injective h,
+    simp only at this,
+    assumption
+  end⟩,
+  inv_fun := λ ⟨⟨f, g⟩, disj⟩,
+    ⟨λ x, match x with
+    | (sum.inl a) := f a
+    | (sum.inr b) := g b
+    end,
+    begin
+      rintros (a₁|b₁) (a₂|b₂) f_eq;
+      simp only [equiv.coe_fn_symm_mk, sum.elim_inl, sum.elim_inr] at f_eq,
+      { rw f.injective f_eq },
+      { simp! only at f_eq, exfalso, exact disj ⟨⟨a₁, by simp⟩, ⟨b₂, by simp [f_eq]⟩⟩  },
+      { simp! only at f_eq, exfalso, exact disj ⟨⟨a₂, by simp⟩, ⟨b₁, by simp [f_eq]⟩⟩  },
+      { rw g.injective f_eq }
+    end⟩,
+  left_inv := λ f, by { dsimp only, ext, cases x; simp! },
+  right_inv := λ ⟨⟨f, g⟩, _⟩, by { simp only [prod.mk.inj_iff], split; ext; simp! } }
+
+/-- Pairs of embeddings with disjoint ranges are equivalent to a dependent sum of embeddings,
+in which the second embedding cannot take values in the range of the first. -/
+def prod_embedding_disjoint_equiv_sigma_embedding_restricted {α β γ : Type*} :
+  {f : (α ↪ γ) × (β ↪ γ) // disjoint (set.range f.1) (set.range f.2)} ≃
+  (Σ f : α ↪ γ, β ↪ ↥((set.range f)ᶜ)) :=
+{ to_fun := λ ⟨⟨f, g⟩, disj⟩, ⟨f, ⟨λ b, ⟨g b, set.disjoint_right.mp disj (⟨b, rfl⟩)⟩,
+  λ _ _ eq, by { rw subtype.mk_eq_mk at eq, exact g.injective eq }⟩⟩,
+  inv_fun := λ ⟨f, g⟩, ⟨⟨f, g.trans (function.embedding.subtype _)⟩,
+  begin
+    dsimp only,
+    have : set.range (g.trans (function.embedding.subtype (λ x, x ∈ (set.range f)ᶜ)))
+      ⊆ (set.range f)ᶜ,
+    { rintros _ ⟨t, rfl⟩, -- `simp` will go down a wrong track on its own
+      simp only [function.embedding.coe_subtype, function.embedding.trans_apply, subtype.coe_prop]},
+    exact set.disjoint_of_subset_right this disjoint_compl_right
+  end⟩,
+  right_inv := λ ⟨_,_⟩, by simp!,
+  left_inv := λ ⟨⟨_,_⟩,_⟩, by { dsimp only, ext; simp! } }
```

Comment:
I was able to split this into chain of operations which are maybe worth having in mathlib somewhere:
```suggestion
/-- A subtype of a `prod` is equivalent to a sigma type whose fibers are subtypes. -/
def subtype_prod_equiv_sigma_subtype {α β : Type*} (p : α → β → Prop) :
  {x : α × β // p x.1 x.2} ≃ Σ a, {b : β // p a b} :=
{ to_fun := λ x, ⟨(x : α × β).1, (x : α × β).2, x.prop⟩,
  inv_fun := λ x, ⟨⟨x.1, x.2⟩, x.2.prop⟩,
  left_inv := λ x, by ext; refl,
  right_inv := λ ⟨a, b, pab⟩, rfl }

/-- Embeddings whose range lies within a set are equivalent to embeddings to that set. -/
def embedding_subtype_range_le_equiv (α : Type*) {β : Type*} (bs : set β) :
  {f : α ↪ β // set.range f ⊆ bs} ≃ (α ↪ bs) :=
{ to_fun := λ f,
  { to_fun := λ a, ⟨f a, f.2 ⟨a, rfl⟩⟩,
    inj' := λ a a' h, (f : α ↪ β).injective $ subtype.ext_iff.1 h },
  inv_fun := λ f, ⟨f.trans (function.embedding.subtype _), λ  _ ⟨a, ha⟩, ha ▸ (f a).prop⟩,
  left_inv := λ x, by { ext, refl },
  right_inv := λ x, by { ext, refl } }

/-- Pairs of embeddings with disjoint ranges are equivalent to a dependent sum of embeddings,
in which the second embedding cannot take values in the range of the first. -/
def prod_embedding_disjoint_equiv_sigma_embedding_restricted {α β γ : Type*} :
  {f : (α ↪ γ) × (β ↪ γ) // disjoint (set.range f.1) (set.range f.2)} ≃
  (Σ f : α ↪ γ, β ↪ ↥((set.range f)ᶜ)) :=
(subtype_prod_equiv_sigma_subtype $
  λ (a : α ↪ γ) (b : β ↪ _), disjoint (set.range a) (set.range b)).trans $
  equiv.sigma_congr_right $ λ a,
    (subtype_equiv_prop begin
      ext f,
      rw set.subset_compl_iff_disjoint,
      exact disjoint.comm.trans disjoint_iff,
    end).trans (embedding_subtype_range_le_equiv _ _)
```

Reply:
this is really amazing! maybe the top one can go in `data.equiv.basic`, the second one I'm torn on whether it should stay here or be moved to be near `function.embedding.cod_restrict`, and the third one should (probably?) stay here

Reply:
feel free to push whatever you think is best by the way - I'm really busy today but would be nice to get this though the cogs of CI